File: moshi-lab/src/test/java/ga/overfullstack/MoshiLab.java
package ga.overfullstack;

import static ga.overfullstack.utils.Utils.readFileFromTestResource;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import com.squareup.moshi.Moshi;
import java.io.IOException;
import java.util.List;
import org.junit.jupiter.api.Test;

class MoshiLab {

  @Test
  void customListAdapter() throws IOException {
    final var beanStr = readFileFromTestResource("bean.json");
    final var moshi = new Moshi.Builder().build();
    final var beanJsonAdapter = moshi.adapter(Bean.class);
    final var bean = beanJsonAdapter.fromJson(beanStr);
    assertNotNull(bean);
    assertEquals("value1", bean.field1);
    assertEquals(2, bean.items.size());
  }

  static class Bean {
    String field1;
    List<String> items;
  }
}

File: moshi-lab/src/test/kotlin/ga/overfullstack/MoshiLab.kt
package ga.overfullstack

import com.squareup.moshi.Moshi
import com.squareup.moshi.adapter
import ga.overfullstack.utils.readFileFromTestResource
import io.kotest.matchers.maps.shouldContainExactly
import org.junit.jupiter.api.Test

internal class MoshiLab {
  @OptIn(ExperimentalStdlibApi::class)
  @Test
  fun readJsonToMap() {
    val mapStr = readFileFromTestResource("obj.json")
    val mapAdapter = Moshi.Builder().build().adapter<Map<String, String>>()
    val map = mapAdapter.fromJson(mapStr)
    println(map)
  }

  @OptIn(ExperimentalStdlibApi::class)
  @Test
  fun readCollectionToListMap() {
    val collectionStr = readFileFromTestResource("collection.json")
    val mapAdapter = Moshi.Builder().build().adapter<Any?>()
    val collection = mapAdapter.fromJson(collectionStr)
    println(collection)
  }

  @OptIn(ExperimentalStdlibApi::class)
  @Test
  fun readJsonToAnyMap() {
    val nestedBeanStr = readFileFromTestResource("nested-bean.json")
    val mapAdapter = Moshi.Builder().build().adapter<Map<*, *>>()
    val any = mapAdapter.fromJson(nestedBeanStr)!! 
    (any["bean"] as Map<String, *>) shouldContainExactly mapOf("name" to "member", "items" to listOf("item1", "item2"))
  }

  @OptIn(ExperimentalStdlibApi::class)
  @Test
  fun readJsonToAny() {
    val nestedBeanStr = readFileFromTestResource("nested-bean.json")
    val anyAdapter = Moshi.Builder().build().adapter<Any>()
    val any = anyAdapter.fromJson(nestedBeanStr)!! // It reads to LinkedHashTreeMap like javascript JSON
    println(any)
  }
}
